sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java

```diff
@@ -760,66 +763,69 @@ private String logsDir(long now) {
-                    .withTopic(options.getKafkaSinkTopic())
-                    .withValueSerializer(ByteArraySerializer.class)
-                    .values());
+    NexmarkUtils.console("Writing events to Kafka Topic %s", options.getKafkaEventsTopic());
+    checkArgument((options.getBootstrapServers() != null), "Missing --bootstrapServers");
 
+    PCollection<byte[]> eventToBytes = events.apply("Event to bytes", ParDo.of(EVENT_TO_BYTEARRAY));
+    eventToBytes.apply(
+        KafkaIO.<Void, byte[]>write()
+            .withBootstrapServers(options.getBootstrapServers())
+            .withTopic(options.getKafkaEventsTopic())
+            .withValueSerializer(ByteArraySerializer.class)
+            .values());
   }
 
-
   static final DoFn<KV<Long, byte[]>, Event> BYTEARRAY_TO_EVENT =
-          new DoFn<KV<Long, byte[]>, Event>() {
-            @ProcessElement
-            public void processElement(ProcessContext c) {
-              byte[] encodedEvent = c.element().getValue();
-              try {
-                Event event = CoderUtils.decodeFromByteArray(Event.CODER, encodedEvent);
-                c.output(event);
-              } catch (CoderException e) {
-                LOG.error("Error while decoding Event from Kafka message: serialization error");
-              }
-            }
-          };
+      new DoFn<KV<Long, byte[]>, Event>() {
+        @ProcessElement
+        public void processElement(ProcessContext c) {
+          byte[] encodedEvent = c.element().getValue();
+          try {
+            Event event = CoderUtils.decodeFromByteArray(Event.CODER, encodedEvent);
+            c.output(event);
+          } catch (CoderException e) {
+            LOG.error("Error while decoding Event from Kafka message: serialization error");
+          }
+        }
+      };
 
   /**
    * Return source of events from Kafka.
    */
-  private PCollection<Event> sourceEventsFromKafka(Pipeline p) {
-    NexmarkUtils.console("Reading events from Kafka Topic %s", options.getKafkaSourceTopic());
-
-    checkArgument(!Strings.isNullOrEmpty(options.getBootstrapServers()),
-        "Missing --bootstrapServers");
+  private PCollection<Event> sourceEventsFromKafka(Pipeline p, final Instant now) {
+    NexmarkUtils.console("Reading events from Kafka Topic %s", options.getKafkaEventsTopic());
+    checkArgument((options.getBootstrapServers() != null), "Missing --bootstrapServers");
 
-    KafkaIO.Read<Long, byte[]> read = KafkaIO.<Long, byte[]>read()
+    KafkaIO.Read<Long, byte[]> read =
+        KafkaIO.<Long, byte[]>read()
             .withBootstrapServers(options.getBootstrapServers())
-            .withTopic(options.getKafkaSourceTopic())
+            .withTopic(options.getKafkaEventsTopic())
             .withKeyDeserializer(LongDeserializer.class)
-            .withValueDeserializer(ByteArrayDeserializer.class);
+            .withValueDeserializer(ByteArrayDeserializer.class)
+            .withStartReadTime(now)
```

Comment:
Why do we want to set this? Is it always supported or does time to offset index need to be enabled for a topic in Kafka?

Reply:
Sounds good. One caveat is that if Kafka server does not support offset index (not sure if that is always on in newer versions), it defaults to reading from latest offset I think. We can leave it on for now.